policyengine_uk/data/datasets/frs/imputations/capital_gains.py

```diff
@@ -18,61 +20,41 @@
-
-
-def impute_capital_gains(total_income: float, age: float) -> float:
-    if total_income < 0 or age < 18:
-        return 0
-    distribution_row = capital_gains[
-        (capital_gains["minimum_total_income"] <= total_income)
-        & (capital_gains["maximum_total_income"] > total_income)
-    ]
-    percent_with_gains = distribution_row["percent_with_gains"].values[0]
-    has_gains = np.random.choice(
-        [0, 1], p=[1 - percent_with_gains, percent_with_gains]
-    )
-    if not has_gains:
-        return 0
-    sample_percentile = np.random.random()
-    for i in range(len(splines)):
-        if lower_income_bounds[i] > total_income:
-            continue
-    i -= 1
-    spline = splines[lower_income_bounds[i]]
-    return spline(sample_percentile) * uprating_from_2017
-
-
-def add_capital_gains_to_dataset(dataset: Dataset, time_period: str):
-    sim = Microsimulation(dataset=dataset)
-    total_income = sim.calculate("total_income", time_period)
-    age = sim.calculate("age", time_period)
-    imputed_gains = []
-    for income, age in tqdm(list(zip(total_income, age))):
-        imputed_gains.append(impute_capital_gains(income, age))
-
-    data = sim.dataset.load_dataset()
-    data["capital_gains"] = {time_period: np.array(imputed_gains)}
-    sim.dataset.save_dataset(data)
-
-
-if __name__ == "__main__":
-    add_capital_gains_to_dataset("enhanced_frs", 2021)
+def impute_capital_gains(time_period: int):
+    """Assumes that the capital gains distribution is the same for all years.
+    """
+
+    sim = Microsimulation(dataset="experimental_enhanced_frs")
+    ti = sim.calculate("total_income", time_period)
+    person_weight = sim.calculate("person_weight", time_period).values
+    has_capital_gains = np.zeros_like(person_weight) > 0
+    first_half = np.concatenate([np.ones(len(person_weight)//2), np.zeros(len(person_weight)//2)]) > 0
+    pred_cg = np.ones_like(~first_half)
+    blend_factor = torch.tensor(np.zeros(first_half.sum()), requires_grad=True)
+
+    def loss(blend_factor):
+        for i in range(len(capital_gains)):
+            np.random.seed(i)
+            lower = capital_gains.minimum_total_income.iloc[i]
+            upper = capital_gains.maximum_total_income.iloc[i]
+            selection = ti.between(lower, upper)
+            first_half_of_selection = selection * first_half
+            second_half_of_selection = selection * ~first_half
+            fraction_with_capital_gains = capital_gains.percent_with_gains.iloc[i]
+            person_weights_subset = person_weight[selection][:selection.sum()//2]
+            person_weight[first_half_of_selection] = person_weights_subset * (1 - fraction_with_capital_gains)
+            person_weight[second_half_of_selection] = person_weights_subset * fraction_with_capital_gains
+            has_capital_gains[second_half_of_selection] = True
+            row = capital_gains.iloc[i]
+            spline = UnivariateSpline(
```

Comment:
To ensure monotonicity, see https://github.com/PolicyEngine/policyengine-uk/issues/816#issuecomment-1962969566
```suggestion
            spline = PchipInterpolator(
```

Reply:
OK so I just tried this and got pretty strange results- capital gains at the high end fell a lot for some reason. We should probably still add it and I'm probably just using it wrong, but needs more careful investigation.